File: BTree.c
#include <stdio.h>
#include <string.h>

int degree;
typedef struct BTreeNode
{
	int *keys; // An array of keys
	struct BTreeNode *P; // An array of parent pointer
	struct BTreeNode **C; // An array of child pointers
	int n; // Current number of Keys
	bool leaf; // Is true when node is leaf
} BTreeNode;

void BTreeInit(int _degree)
{
	root = NULL;
	degree = _degree - 1;
}

BTreeNode* createNode(bool _leaf)
{
	BTreeNode* newNode = (BTreeNode*)malloc(sizeof(BTreeNode));
	int i;

	newNode->leaf = _leaf;

	newNode->keys = (int*)malloc(sizeof(int)*(degree+1));
	newNode->C = (BTreeNode**)malloc(sizeof(BTreeNode*)*(degree+2));
	
	for(i=0; i<degree+2; i++)
		newNode->C[i] = NULL;
	
	newNode->n = 0;
	newNode->P = NULL;
	
	return newNode;
}

void traverse(BTreeNode* present)
{
	if (root != NULL)
	{
		int i;
		for(i=0; i<present->n; i++)
		{
			if (present->leaf == false)
				traverse(present->C[i]);
			printf(" ");
			printf("%d", present->keys[i]);
		}
		
		if (present->leaf == false)
			traverse(present-C[i]);
	}
}

BTreeNode* search(int k)
{
	return (root==NULL)?NULL:_search(root, k);
}

BTreeNode* _search(BTreeNode* present, int k)
{
	int i=0;
	while(i<present->n && k > present->keys[i])
		i++;

	if(present->keys[i] == k)
		return present;

	if(present->leaf == true)
		return NULL;

	return _search(present->C[i], k);
}

void insertElement(int k)
{
	if (search(k) != NULL)
	{
		printf("The tree already has %d \n", k);
		return;
	}

	if (root == NULL)
	{
		root = createNode(true);
		root->P = NULL;
		root->keys[0] = k;
		root->n = 1;
	}
	else
		_insert(root, k);
}

void _insert(BTreeNode *present, int k)
{
	int i=present->n;
	
	if(present->leaf == true)
	{
		while(i>0 && (present->keys[i-1]) > k)
		{
			present->keys[i] = present->keys[i-1];
			i--;
		}
		present->keys[i] = k;
		(present->n)++;
		_balancing(present);
	}
	else
	{
		while(i>0 && (present->keys[i-1] > k))
		{
			i--;
		}
		
		_insert(present->C[i], k);
	}
}

void _balancing(BTreeNode* present)
{
	BTreeNode* parent;
	
	if (present->n <= t)
		return;

	else if (present->P == NULL)
	{
		root = _splitChild (present);
		return;
	}
	
	else
	{
		parent = _splitChild(present);
		_balancing(parent);
	}
}

BTreeNode* _splitChild(BTreeNode* present)
{
	int i;
	int splitIndex;
	int risingKey;
	int parentIndex;
	
	BTreeNode* currentParent;
	BTreeNode* left;
	BTreeNode* right = _createNode(present->leaf);
	
	splitIndex = t/2;

	right->n = present->n - splitIndex - 1;
	risingKey = present->keys[splitIndex];

	if(present->P != NULL)
	{
		currentParent = present->P;
		for(parentIndex = 0; parentIndex < currentParent->n + 1 && currentParent->C[parentIndex] != present; parentIndex++);
		for(i=currentParent->n; i > parentIndex; i--)
		{
			currentParent->C[i+1] = currentParent->C[i];
			currentParent->keys[i] = currentParent->keys[i-1];
		}
		currentParent->n = currentParent->n + 1;
		currentParent->keys[parentIndex] = risingKey;
		
		currentParent->C[parentIndex + 1] = right;
		right->P = currentParent;
	}
	
	for(i=splitIndex + 1; i<present->n + 1; i++)
	{
		right->C[i - spliIndex - 1] = present->C[i];
		if (present->C[i] != NULL)
		{
			right->leaf = false;
			
			if(present->C[i] != NULL)
			{
				present->C[i]->P = right;
			}
			present->C[i] = NULL;
		}
	}

	for(i=splitIndex + 1; i < present->n; i++)
		right->keys[i - splitIndex - 1] = present->keys[i];
	
	left = present;
	left->n = splitIndex;

	if (present->P != NULL)
		return present->P;
	else
	{
		root = _createNode(present->leaf);
		root->keys[0] = risingKey;
		root->n = 1;
		root->C[0] = left;
		root->C[1] = right;
		left->P = root;
		right->P = root;
		root->leaf = false;
		return root;
	}
}


void removeElement(int k)
{
	if(!root)
	{
		printf("The tree is empty\n");
		return;
	}

	_remove(root, k);

	if (root->n == 0)
	{
		BTreeNode *tmp = root;
		if(root->leaf)
			root = NULL;
		else
		{
			root = root->C[0];
			root->P = NULL;
		}
		free(tmp);
	}
	return;
}


void _remove(BTreeNode* present, int k)
{
	BTreeNode* del_node = search(k)
	if(del_node == NULL)
	{
		printf("%d is not exist\n", k);
		return;
	}

	int child_num = del_node->n;
	int i = del_node->n;
	i--;
	int temp = del_node->keys[i];
	if(del_node->leaf)
	{
		while(i>0 && temp != k)
		{
			del_node->keys[i-1] = temp;
			temp = del_node->keys[i-1];
			i--;
		}
		(del_node->n)--;
		_balancingAfterDel(del_node);
	}
	

	else
	{
		while(i>0 && temp!=k)
			i--;
		del_node->keys[i] = del_node->C[i]->keys[del_node->C[i]->n-1];
		(del_node->C[i]->n)--;
		_balancingAfterDel(del_node->C[i]);
	}
	
	return;
}

void _balancingAfterDel(BTreeNode* present)
{
	int minKeys = (degree+2)/2 - 1;
	BTreeNode* parent;
	BTreeNode* next;
	int parentIndex = 0;
	
	if(present->n < minKeys)
	{	
		if(present->P == NULL)
		{
			if(present->n == 0)
			{
				root = present->C[0];
				if(root !=NULL)
					root->P = NULL;
			}
		}
		else
		{
			parent = present->P;
			for(parentIndex=0; parent->C[parentIndex] != present; parentIndex++);
			if(parentIndex>0 && parent->C[parentIndex-1]->n > minKeys)
			{
				_borrowFromLeft(present, parentIndex);
			}
			else if(parentIndex < parent->n && parent->C[parentIndex+1]->n > minKeys)
			{
				_borrowFromRight(present, parentIndex);
			}
			else if(parentIndex==0)
			{
				next = _merge(present);
				_balancingAfterDel(next->P);
			}
		}
	}
}

void _borrowFromRight(BTreeNode* present, int parentIdx)
{
	int i;
	BTreeNode* rightSib;
	BTreeNode* parentNode = present->P;

	rightSib = parentNode->C[parentIdx + 1];
	present->n++;
	
	present->keys[present->n - 1] = parentNode->keys[parentIdx];
	parentNode->keys[parentIdx] = rightSib->keys[0];

	if(!present->leaf)
	{
		present->C[present->n] = rightSib->C[0];
		presnet->C[present->n]->P = present;

		for(i=1; i<rightSib->n + 1; i++)
			rightSib->C[i-1] = rightSib->C[i];
	}
	
	for(i=1; i<rightSib->n; i++)
		rightSib->keys[i-1] = rightSib->keys[i];
	
	rightSib->n--;
}

void _borrowFromLeft(BTreeNode* present, int parentIdx)
{
	int i;
	BTreeNode* leftSib;
	BTreeNode* parentNode = present->P;
	present->n = present->n + 1;

	for(i=present->n -1; i>0; i--)
		present->keys[i] = present->C[i-1];
	leftSib = parentNode->C[parentIdx-1];

	if(!present->leaf)
	{
		for(i=present->n; i>0; i--)
			present->C[i] = present->C[i-1];
	
		present->C[0] = leftSib->C[leftSib->n];
		leftSib->C[leftSib->n] = NULL;
		present->C[0]->P = present;
	}

	present->keys[0] = parentNode->keys[parentIdx - 1];
	parentNode->keys[parentIdx-1] = leftSib->keys[leftSib->n -1];
	
	leftSib->n = leftSib->n - 1;
}

BTreeNode* _merge(BTreeNode* present)
{
	BTreeNode* parentNode = present->P;
	int parentIndex = 0;
	int fromParentIndex;
	int i;
	
	for(parentIndex = 0; parentNode->C[parentIndex] != present; parentIndex++);
	BTreeNode* rightSib = parentNode->C[parentIndex+1];

	present->keys[present->n] = parentNode->keys[parentIndex];
	fromParentIndex = present->n;
	
	for(i=0; i<rightSib->n;i++)
		present->keys[present->n +1 +i] = rightSib->keys[i];

	if(!present->leaf)
		for(i=0; i<=rightSib->n; i++)
		{
			present->C[present->n +1+i] = rightSib->C[i];
			present->C[present->n +1+i]->P = present;
		}
	
	for(i=parentIndex+1; i<parentNode->n; i++)
	{
		parentNode->C[i] = parentNode->C[i+1];
		parentNode->keys[i-1] = parentNode->keys[i];
	}
	parentNode->n--;
	present->n = present->n + rightSib->n + 1;
	return present;
}

int _getLevel(BTreeNode* present)
{
	int i;
	int maxLevel = 0;
	int temp;
	if(present==NULL)
		return maxLevel;
	if(present->leaf == ture)
		return maxLevel++;
	
	for(i=0; i<present->n+1; i++)
	{
		temp = _getLevel(present->C[i]);
		
		if(temp>maxLevel)
			maxLevel = temp;
	}
	
	return maxLevel + 1;
}

void _getNumberOfNodes(BTreeNode* present, int* numNodes, int level)
{
	int i;
	if(present==NULL)
		return;
	
	if(present->leaf==false)
	{
		for(i=0; i<present->n+1; i++)
			_getNumberOfNodes(present->C[i], numNodes, level + 1);
	}
	numNodes[level] += 1;
}

void _mappingNodes(BTreeNode* present, BTreeNode ***nodePtr, int* numNodes, int level)
{
	int i;
	if(present==NULL)
		return;
	
	if(present->leaf == false)
	{
		for(i=0; i<present->n+1; i++)
			_mappingNodes(present->C[i], nodePtr, numNodes, level+1);
	}

	nodePtr[level][numNodes[level]] = present;
	numNodes[level] += 1;
}

void printTree()
{
	int level;
	int* numNodes;
	int i,j,k;

	level = _getLevel(root);
	numNodes = (int*) malloc(sizeof(int)* level);
	memset(numNodes, 0,0 level*sizeof(int));
}

int main()
{
	BTreeNode temp;
	int arr[3] = {1, 2, 3};
	temp.keys = arr;
	temp.leaf = False;
	temp.n = strlen(arr);
	return 0;
}

File: Python/BTree/BTree.py
class BTreeNode(object):
	def __init__(self, degree, leaf):
		#leaf is boolean variable that means cur_node is leaf or not
		#현 node가 leaf인지 아닌지에 대한 변수
		self.leaf = leaf

		#keys that cur_node has
		#현 node가 가지는 keys
 		self.keys = []
		
		#childs that cur_node has
		#현 node가 가지는 childs
		self.childs = []

		#parent of cur_node
		#현 node의 parent
		self.parent = None

		#degree of BTree
		#BTree의 degree
		self.degree = degree

class BTree(object):
	def __init__(self, degree):
		self.degree = degree - 1
		self.root = None

	def traverse(self):
		if len(self.root.keys) != 0:
			_traverse(self.root)

	def _traverse(cur_node):
		for i in range(len(cur_node.keys)):
			if not cur_node.leaf:
				_traverse(cur_node.childs[i])
			
			print(" ");
			print(cur_node.keys[i])

		if not cur_node.leaf:
			traverse(cur_node.childs[len(cur_node.keys)]

	def search(self, key):
		if len(self.root.keys) == 0:
			return None
		else:
			return _search(self.root, key)

	def _search(cur_node, key):
		i=0
		while i<len(cur_node.keys) and key > cur_node.keys[i]:
			i += 1

		if cur_node.keys[i] == key
			return cur_node

		if cur_node.leaf:
			return None

		return _search(cur_node.childs[i], key)

	def insertElement(self, key):
		if search(key) != None:
			print("The tree already has %d \n", key)
			return None

		if self.root == None:
			self.root = BTreeNode(self.degree, True)
			self.root.keys.append(key)

		else:
			_insert(self.root, key)

	def _insert(cur_node, key):
		i = len(cur_node.keys)

		if cur_node.leaf:
			while i>0 and cur_node.keys[i-1] > key:
				cur_node.keys[i] = cur_node.keys[i-1]
				i -= 1

			cur_node.keys[i] = key
			_balancing(cur_node)

		else:
			while i>0 and cur_node.keys[i-1] > key:
				i -= 1
			
			_insert(cur_node.childs[i], key)

	def _balancing(self, cur_node):
		if len(cur_node.keys) <= degree:
			return None
		
		elif not cur_node.parent:
			self.root = _splitChild(cur_node)
			return None

		else:
			parent = _splitChild(cur_node)
			_balancing(parent)

	def _splitChild(self, cur_node):
		currentParent = None
		splitIndex = self.degree / 2
		risingKey = cur_node.keys[splitIndex]
		right = BTreeNode(self.degree, cur_node.leaf)
		if cur_node.parent:
			currentParent = cur_node.parent
			i = 0
			while i < len(currentParent)+1 and currentParent.child[i] != cur_node:
				i += 1
			j = len(currentParent.keys)
			while j > i:
				currentParent.child[j+1] = currentParent.child[j]
				currentParent.keys[j] = currentParent.keys[j-1]
				j -= 1

			currentParent.keys[i] = risingKey
			currentParent.child[j+1] = right
			right.parent = currentParent

		i = splitIndex + 1
		while  i < (len(cur_node.keys)+1):
			right.child[i - splitIndex - 1]  = cur_node.child[i]
			if cur_node.child[i]:
				right.leaf = False
				if cur_node.child[i]:
					cur_node.child[i].parent = right
				cur_node.child[i] = None
			
		i = splitIndex + 1
		while i< len(cur_node.keys):
			right.keys[i - splitIndex - 1] = cur_node.keys[i]
		
		left = cur_node
		
		if cur_node.parent:
			return cur_node.parent
		
		else:
			self.root = BTreeNode(self.degree, cur_node.leaf)
			self.root.keys[0] = risingKey
			self.child[0] = left
			self.child[1] = right
			left.parent = self.root
			right.parent = self.root
			self.root.leaf = False
			return self.root
	
	def removeElement(self, key):
		if not self.root:
			print("The tree is empty\n")
			return
			
		_remove(self.root, key)
		
		if len(self.root.keys) == 0:
			tmp = self.root
			if root.leaf:
				self.root = None
			else
				self.root = self.root.child[0]
				self.root.parent = None
		
		return None
	
	def _remove(cur_node, key):
		del_node = search(key)
		if del_node == None:
			print("%d is not exist\n", key)
			return None
		
		i = len(del_node.keys) - 1
		temp = del_node.keys[i]
		
		if del_node.leaf:
			while i>0 and temp != key:
				del_node.keys[i-1] = temp
				temp = del_node.keys[i-1]
				i -= 1
			
			_balancingAfterDel(del_node)
		
		else:
			while i>0 and temp!=key:
				i -= 1
			
			del_node.keys[i] = del_node.child[i].keys[len(del_node.child[i]) - 1]
			_balancingAfterDel(del_node.child[i])
		
		return None
		
	def _balancingAfterDel(self, cur_node):
		minKeys = (degree+2)/2 - 1
		parentIndex = 0
		
		if len(cur_node.keys) < minKeys:
			if not cur_node.parent:
				if len(cur_node.keys) == 0:
					self.root = cur_node.child[0]
					if self.root:
						self.root.parent = None
			else:
				parent = cur_node.parent
				while parent.child[parentIndex] != cur_node:
					parentIndex += 1
				if parentIndex > 0 and len(parent.child[parentIndex-1].keys) > minKeys:
					_borrowFromLeft(cur_node, parentIndex)
				elif parentIndex < len(parent.keys) and len(parent.child[parentIndex+1].keys) > minKeys:
					_borrowFromRight(cur_node, parentIndex)
				elif parentIndex==0:
					next = _merge(cur_node)
					_balancingAfterDel(next.parent)
	
	def _borrowFromRight(cur_node, parentIdx):
		parentNode = cur_node.parent
		
		rightSib = parentNode.child[parentIdx + 1]
		present->keys[len(cur_node.keys) - 1] = parentNode.keys[parentIdx]
		parentNode.keys[parentIdx] = rightSib.keys[0]
		
		
		if not cur_node.leaf:
			cur_node.child[len(present.keys)] = rightSib.child[0]
			cur_node.child[len(present.keys)] = cur_node
			
			for i in range(len(rightSib.keys) + 1):
				rightSib.child[i-1] = rightSib.child[i]
				
		for i in range(len(rightSib.keys)):
			rightSib.keys[i-1] = rightSib.keys[i]
	
	def _borrowFromLeft(cur_node, parentIdx):
		parentNode = cur_node.parent
		
		i = len(cur_node.keys) - 1
		while i > 0:
			cur_node.keys[i] = cur_node.keys[i-1]
			i--
		leftSib = parentNode.child[parentIdx - 1]
		
		if not cur_node.leaf:
			i = len(cur_node.keys)
			while i > 0:
				cur_node.child[i] = cur_node.child[i-1]
				i--
			
			cur_node.child[0] = leftSib.child[len(leftSib.keys)]
			leftSib.child[len(leftSib.keys)] = None
			cur_node.child[0].parent = cur_node
		
		cur_node.keys[0] = parentNode.keys[parentIdx-1]
		parentNode.keys[parentIdx-1] = leftSib.keys[len(leftSib.keys) - 1]
		
	def _merge(cur_node):
		parentNode = cur_node.parent
		parentIndex = 0
		
		while parentNode.child[parentIndex] != cur_node:
			parentIndex += 1
		
		rightSib = parentNode.child[parentIndex + 1]
		
		cur_node.keys[len(cur_node.keys)] = parentNode.keys[parentIndex]
		fromParentIndex = len(present.keys)
		
		for i in range(len(rightSib.keys)):
			cur_node.keys[len(cur_node) + 1 + i] = rightSib.keys[i]
		
		if not cur_node.leaf:
			for i in range(len(rightSib.keys)):
				cur_node.child[len(cur_node) + 1 + i] = rightSib.child[i]
				cur_node.child[len(cur_node) + 1 + i] = cur_node
			
		i = parentIndex + 1
		while i < len(parentNode.keys):
			parentNode.child[i] = parentNode.child[i + 1]
			parentNdoe.keys[i-1] = parentNode.keys[i]
			
		return cur_node
		
	
